internal/gcsx/client_readers/gcs_reader_test.go

```diff
@@ -0,0 +1,264 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcsx
+
+import (
+	"context"
+	"io"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/v2/common"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/clock"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/gcsx"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/storage"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/storage/fake"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/storage/gcs"
+	testUtil "github.com/googlecloudplatform/gcsfuse/v2/internal/util"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+////////////////////////////////////////////////////////////////////////
+// Helpers
+////////////////////////////////////////////////////////////////////////
+
+func (t *gcsReaderTest) readAt(offset int64, size int64) (gcsx.ReaderResponse, error) {
+	t.gcsReader.CheckInvariants()
+	defer t.gcsReader.CheckInvariants()
+	return t.gcsReader.ReadAt(t.ctx, make([]byte, size), offset)
+}
+
+////////////////////////////////////////////////////////////////////////
+// Boilerplate
+////////////////////////////////////////////////////////////////////////
+
+type gcsReaderTest struct {
+	suite.Suite
+	ctx        context.Context
+	object     *gcs.MinObject
+	mockBucket *storage.TestifyMockBucket
+	gcsReader  *GCSReader
+}
+
+func TestGCSReaderTestSuite(t *testing.T) {
+	suite.Run(t, new(gcsReaderTest))
+}
+
+func (t *gcsReaderTest) SetupTest() {
+	t.object = &gcs.MinObject{
+		Name:       testObject,
+		Size:       17,
+		Generation: 1234,
+	}
+	t.mockBucket = new(storage.TestifyMockBucket)
+	t.gcsReader = NewGCSReader(t.object, t.mockBucket, common.NewNoopMetrics(), nil, 200)
+	t.ctx = context.Background()
+}
+
+func (t *gcsReaderTest) TearDownTest() {
+	t.gcsReader.Destroy()
+}
+
+////////////////////////////////////////////////////////////////////////
+// Tests
+////////////////////////////////////////////////////////////////////////
+
+func (t *gcsReaderTest) Test_NewGCSReader() {
+	// The setup instantiates gcsReader with NewGCSReader.
+	assert.Equal(t.T(), t.object, t.gcsReader.object)
+	assert.Equal(t.T(), t.mockBucket, t.gcsReader.bucket)
+	assert.Equal(t.T(), testUtil.Sequential, t.gcsReader.readType)
+}
```

Comment:
The use of implicit setup via SetupTest obfuscates the purpose of this test. Please refer to http://go/unit-testing-practices?polyglot=go#setUp and amend accordingly.